config/opal_config_asm.m4

```diff
@@ -178,15 +178,38 @@ int main(int argc, char** argv)
 dnl ------------------------------------------------------------------
 
 dnl
-dnl Check to see if a specific function is linkable.
+dnl Helper to Check if a specific function is usable.
 dnl
-dnl Check with:
+dnl
+dnl $1: function name to print
+dnl $2: program to test
+dnl $3: action if success
+dnl #4: action if fail
+dnl
+AC_DEFUN([_OPAL_ASM_CHECK_ATOMIC_FUNC],[
+    AC_LINK_IFELSE([$2],
+        [AC_MSG_RESULT([yes])
+         dnl make sure it works
+         AC_MSG_CHECKING([if $1() gives correct results])
+         AC_RUN_IFELSE([$2],
+               [$3
+                AC_MSG_RESULT([yes])],
+               [$4
+                AC_MSG_RESULT([no])],
+               [AC_MSG_RESULT([cannot test -- assume yes (cross compiling)])])],
```

Comment:
This isn't right; you need to either expand `$3` or `$4` here.  Based on the result message, I assume you wanted to expand `$3`.

There's no great answer for what to do in these situations, but you should definitely make the cache work properly for this check, so that if there's a problem, cross-compiling users can pre-populate a cache file and control their own destiny.

Reply:
@bwbarrett good catch! I added the missing`$3`. I will address the caching later: this macro is invoked in multiple places, so I really need to think that through.

@jsquyres FWIW, yes this PR does use `AC_RUN_IFELSE()` but
 - so does the current code
 - test is assumed to success when cross-compiling.

Reply:
I'm ok with this patch, but we really do need to solve the caching problem before committing.